Got rid of +1 to gravity_distance_scale

ryanpeach opened the pull request:
Reference: #1932

The +1 in the gravity_distance scale formula breaks any ability to convert the point gravity into standard gravitational form.
This also causes precession in orbits.

The gravitation parameter `mu=G*M`
	
Newtons Equation using the gravitational parameter is:
`F = mu*mass/distance^2` [1]
	
The equation in the godot code is:
	`F = gravity * mass / (distance * gravity_distance_scale + 1)^2` [2]

The resultant equation is:
	`mu = distance^2*gravity/(1+distance*gravity_distance_scale)^2` [3]

But this still contains distance, which is weird.
	
So now we can get this if we remove that +1:
	`mu = gravity/gravity_distance_scale^2` [4]

## References

1. https://en.wikipedia.org/wiki/Standard_gravitational_parameter
2. https://github.com/Faless/godot/blob/590afbcac461f87b05391996ca85d32627c81a75/servers/physics_2d/body_2d_sw.cpp
3. https://www.wolframalpha.com/input/?i=u+*+m+%2F+r%5E2+%3D+g+*+m+%2F+%28r+*+s+%2B+1%29%5E2+solve+for+u
4. https://www.wolframalpha.com/input/?i=u+*+m+%2F+r%5E2+%3D+g+*+m+%2F+%28r+*+s%29%5E2+solve+for+u

Commit:
Got rid of +1 to gravity_distance_scale

Commit:
Fixing build errors

Commit:
Misnamed real_t

Commit:
There is no reason to normalize a length 0 vector. May cause a divide by 0

groud commented:
cc @faless

ryanpeach commented:
So, INF's are an inevitability in this case but good catch, but that is because the radius can always be near zero and or the distance scale, and I think those are both on the user ultimately to avoid (that's simple gravity, black holes and the like).

These are some of my proposals:

1. We can max the point gravity to some number like MAX_FLOAT.
2. We can print a debug warning to the console on the existence of INF and then not add INF to gravity.

ryanpeach commented:
The reason you got a number less than FLT_MIN is described here, but I had to look it up:

ryanpeach commented:
I'm running this on a branch of mine and it's working pretty well. Except that I'm having trouble reducing gravity_distance_scale to any number less than 0.0001, but not 0, which is pretty necessary. I'm also encountering the consequences of the infinities, so I'll work on that.

Faless commented:
> So, INF's are an inevitability in this case but good catch, but that is because the radius can always be near zero and or the distance scale, and I think those are both on the user ultimately to avoid (that's simple gravity, black holes and the like).

Well, yes and no, meaning, of course the formula can cause that. but we don't really know if it even make sense in real physics (and the formula itself is just an approximation, not the real deal).
AFAIK the central point of a black hole is a singularity, a point where "we don't know what happens", so it might not be the case.

In general, we MUST deal with that case in a way that doesn't break the whole physics engine it it happens (e.g. devise a workaround so that if it happens, we ignore those values).

> We can print a debug warning to the console on the existence of INF and then not add INF to gravity.

Printing might not be the best solution, as it might cause quite a significant FPS drop if the print is frequent.

> Except that I'm having trouble reducing gravity_distance_scale to any number less than 0.0001

IIRC, the editor does not allow you to set a value smaller than `0.0001`, but you should be able to set it via code.

Calinou commented:
> Printing might not be the best solution, as it might cause quite a significant FPS drop if the print is frequent.

We can use the `WARN_PRINT_ONCE()` macro for this. It can only print once per session.

ryanpeach commented:
So the "safest" alternative is to invent a new parameter, the "event horizon" which is a minimum distance within which the gravity can be constant. Then the user can set this and gravity_distance_scale so low that it's meaningless if they so choose. But since those are constants any physics breaking would be their own fault and we can just print a warning at launch.

ryanpeach commented:
I suppose we could also just force a collision2d boundary around any point gravity areas so that they bounce off or make that the event horizon. Just idea generation at this point.

Commit:
Event Horizon added

ryanpeach commented:
Also, I'm a new C++ developer (7y python). Does putting const everywhere improve performance at all or is it just a protective modifier? I'm used to rust and other similar languages where things are immutable by default.

Faless commented:
using const when things are meant to be constant is okay. I would do the comparison with the event horizon after the multiplication for the distance scale.

ryanpeach commented:
Like this?

IMO I don't think so, I think it should be an absolute pixel radius so it's easier for the user to understand and control. gravity_distance_scale is more of a gravitational fall off rate. I think they should be independent.

```cpp
		const real_t EVENT_HORIZON = 1.0; // TODO: Make user modifiable
			Vector2 v = p_area->get_transform().xform(p_area->get_gravity_vector()) - get_transform().get_origin();			const real_t gravity_distance_scale = p_area->get_gravity_distance_scale();
			gravity += v.normalized() * (p_area->get_gravity() / Math::pow(v.length() * p_area->get_gravity_distance_scale() + 1, 2));			Vector2 v = p_area->get_transform().xform(p_area->get_gravity_vector()) - get_transform().get_origin();
		real_t v_length = v.length();
                 real_t r = v_length * gravity_distance_scale;
		if (gravity_distance_scale > 0 && r > EVENT_HORIZON) {
			real_t new_gravity = (p_area->get_gravity() / Math::pow(r, 2));
			if (isinf(new_gravity)) {
				WARN_PRINT_ONCE("Gravity is Infinite");
			} else {
				gravity += v.normalized() * new_gravity;
			}
		} else {			} else {
			gravity += (p_area->get_transform().xform(p_area->get_gravity_vector()) - get_transform().get_origin()).normalized() * p_area->get_gravity();				gravity += v.normalized() * p_area->get_gravity();
		}			}
```

Faless commented:
The problem is that given a certain event horizon value, it will still
break for a low enough distance scale. The distance scale is meant to be a
way to scale down the base unit value, instead of having 1px = 1m you might
want 1px = 10m, or 1px = 0.1m and so on. For this reason, the event horizon
value, should be expressed in the desired base unit after applying the
scale, and not in pixels.

On Mon, Aug 24, 2020, 20:07 Ryan Peach <notifications@github.com> wrote:

> Like this?
>
> IMO I don't think so, I think it should be an absolute pixel radius so
> it's easier for the user to understand and control. gravity_distance_scale
> is more of a gravitational fall off rate. I think they should be
> independent.
>
> 		const real_t EVENT_HORIZON = 1.0; // TODO: Make user modifiable
> 			Vector2 v = p_area->get_transform().xform(p_area->get_gravity_vector()) - get_transform().get_origin();			const real_t gravity_distance_scale = p_area->get_gravity_distance_scale();
> 			gravity += v.normalized() * (p_area->get_gravity() / Math::pow(v.length() * p_area->get_gravity_distance_scale() + 1, 2));			Vector2 v = p_area->get_transform().xform(p_area->get_gravity_vector()) - get_transform().get_origin();
> 		real_t v_length = v.length();
>                  real_t r = v_length * gravity_distance_scale;
> 		if (gravity_distance_scale > 0 && r > EVENT_HORIZON) {
> 			real_t new_gravity = (p_area->get_gravity() / Math::pow(r, 2));
> 			if (isinf(new_gravity)) {
> 				WARN_PRINT_ONCE("Gravity is Infinite");
> 			} else {
> 				gravity += v.normalized() * new_gravity;
> 			}
> 		} else {			} else {
> 			gravity += (p_area->get_transform().xform(p_area->get_gravity_vector()) - get_transform().get_origin()).normalized() * p_area->get_gravity();				gravity += v.normalized() * p_area->get_gravity();
> 		}			}
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/41386#issuecomment-679310794>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAM4C3VUDP4CZJ5ZL34FT7LSCK3ABANCNFSM4QFP77OQ>
> .
>

ryanpeach commented:
That is true, but I didn't mean that it necessarily has to be in pixels. gravity_distance_scale is not a conversion factor from m to px to whatever, it's an extension of the gravitational constant. There are worlds you may want to create where the unit is lets say in meters, but the gravity doesn't degenerate as fast as in nature, or degenerates faster. gravity_distance_scale is not therefore a conversion factor, but rather an intensity metric. Thus EVENT_HORIZON should be seperate, represented in simple units, not units * gravity_distance_scale, because gravity_distance_scale is not a conversion medium.

ryanpeach commented:
Imagine that EVENT_HORIZON could be drawn as a circle in the editor. When the user then modifies gravity_distance_scale, would you want it to change? I don't think that's what the user would want.

ryanpeach commented:
> The problem is that given a certain event horizon value, it will still
break for a low enough distance scale. 

If a user sets a low enough event horizon that they get undefined behavior, they will get a warning and undesireable effects. But that wont be default, and by default it won't throw INF's the way it is written. But knowing what they are doing, they should be able to not use the event horizion if they so choose. I'm making a very realistic 2D space sim. I will probably set the event horizon to 0, I really don't need it, because I have collision 2d objects surrounding my gravitational points. But I definitely wouldn't want it dynamically recalculated based on gravity_distance_scale, because that doesn't effect the shape of any other thing in my editor. It doesn't change the size of my planets for instance, what if the event horizon suddenly grew outside my planet just because I changed gravity_distance_scale?

ryanpeach commented:
I suppose mathematically I'd just refer you back to `mu = gravity/gravity_distance_scale^2`. So gravity and gravity_distance_scale are really just standins for the `mu` in newtons gravity. The editor doesn't enforce how you see or use these parameters, and there is no universal "distance_scale" in the editor that this parameter effects, so i'd leave it's interpretation as an exercise for the user. We would probably want to unify these parameters into a proper G and mass, except that would break compatability, and also gravity_distance_scale prevents infs in the pow(v_length * gravity_distance_scale, 2.0) operation.

Commit:
Added more protections against INF

ryanpeach commented:
I've added some more protections against getting an INF. No matter what the event horizon, there will never be an INF gravity. Just very high. Note that in the current implementation, there can still be things "zooming off" into oblivion, even with the +1, depending on how strong gravity is (I've done it in a scene already), so in this state there is no functional difference between what exists.

My main argument against the gravity_distance_scale * EVENT_HORIZON is this:

> But I definitely wouldn't want it dynamically recalculated based on gravity_distance_scale, because that doesn't effect the shape of any other thing in my editor. It doesn't change the size of my planets for instance, what if the event horizon suddenly grew outside my planet just because I changed gravity_distance_scale?

So simple usability.

We should work to make EVENT_HORIZON visible in the area2d object and modifiable by user click and drag. As well as giving it setters and getters.

Commit:
Fixed compilation error.

Faless commented:
The `gravity distance scale` is a parameter used to scale distances related to gravity. Which is why the event horizon (which only make sense when the gravity formulas are used) should be using that value IMO.

ryanpeach commented:
If we can get another vote on the matter that would be ideal I think, but from a usability standpoint I think that's inconsistent. There is no universal `distance_scale` in godot, and I don't really like coupled parameters. I should be able to set an event horizon in "units" and expect it to stay where I drew it.

In truth though, do we really need it now? I don't think so.

Commit:
Removed EVENT_HORIZON.

ryanpeach commented:
Removing EVENT_HORIZON avoids an extra parameter to present the user, allows them to use existing builtins like Collision's and higher priority Area's as their "event horizon". The INF solution was solved by never adding INF to gravity and never allowing gravity + x to equal INF. Things will still "zoom away" but always have, and will no longer precess as the +1 solution causes. NaN's are not possible as divide by 0 is checked as well.

ryanpeach commented:
The build for windows failed but I don't think theres a real reason...

madmiraal reviewed (changes requested):
I don't think this change is a good idea. While I appreciate the desire for physically accurate physics, physically accurate physics is not intuitive to users. (e.g. when damping is physically accurate: #19182). Similarly, I expect this change to confuse users.

Intuitively, gravity at or near the `Gravity Point` is not expected to be infinite, it's expected to be equal to `Gravity`. With any `Gravity Distance Scale` > 0, gravity at the `Gravity Point` is still expected to be equal to `Gravity`; and then drop off with distance at a rate proportional to the `Gravity Distance Scale`. However, with this change, gravity near the `Gravity Point` would now be nearly infinite (regardless of whether or not a warning is raised).

Most importantly, with this change, there is no way to achieve the current behaviour. It would either by equal to `Gravity` and not vary with distance, or it would be infinite at the `Gravity Point` and only finally equal `Gravity` at a distance equal to the `Gravity Scale` before dropping off as it did before.

If physically accurate gravity is required, the user can always override physics with the `Custom Integrator`; including taking responsibility for ensuring the attractor has a volume. The lack of a volume being the real reason the gravity around a `Gravity Point` cannot be physically accurate.

Finally, even if this change is still desired, it should be updated to include Bullet physics too.

madmiraal commented:
> This also causes precession in orbits.

The +1 is not the cause of the precession, because it happens with the gravity_distance_scale set to 0 too i.e. when the formula this PR changes is not used.

ryanpeach commented:
> While I appreciate the desire for physically accurate physics, physically accurate physics is not intuitive to users. (e.g. when damping is physically accurate: #19182). Similarly, I expect this change to confuse users.

I don't want to put a lot of extra work into this PR as it works fine for me on my fork. However, IMO, this statement is simply similar to saying "our game engine is not interested in supporting the space-sim genre"

ryanpeach commented:
Do note that both Unreal and Unity have physically accurate point gravity models. And, as all physically accurate point gravity models do, there is a singularity at the center. You just have to put a wall around it. I don't think that the users are confused by accuracy, I think they are confused by arbitrary physics systems that are fake. Instead of making fake the default, lets make real the default, and make fake the thing that requires custom integrators, like most game engines do.

ryanpeach commented:
> The +1 is not the cause of the precession, because it happens with the gravity_distance_scale set to 0 too i.e. when the formula this PR changes is not used.

Can you demo this? On my branch, precession is gone.

ryanpeach commented:
Reviewing  #19182, the community did not choose to use unrealistic physics with dampening. They chose to use a different arbitrary scale for a dampening parameter. This is a good choice. We can make such choices. But denying this change is paramount to actually choosing to use an equation that is fake, which I'm not aware of any reputable physics engine choosing to do.

Calinou commented:
@ryanpeach Please use the Edit button in the top-right corner of your comment (in the **…** dropdown) instead of multi-posting.

Faless commented:
> The +1 is not the cause of the precession, because it happens with the gravity_distance_scale set to 0 too i.e. when the formula this PR changes is not used.

Yeah but the formula without the distance scale is supposed to cause the precession since it applies a fixed force.
Being the one that implemented the gravity distance scale `+1` "hack" I agree it's wrong, and it felt wrong at the time too (I think it was my first or second contribution to Godot, and didn't know the consequences :) ).

In the end, for my game, I went back to use `Custom Integrator` as I noticed the precession, but I actually ended up using the correct formula as in this PR and the precession was infinitesimally smaller.
You can't totally get rid of it (due to normal integration errors), but you can achieve stable orbits (wobbling if you do Newton prediction, but stable without a noticeable precession) for minutes/hours, by using sensible values. I think the reason is that the `+1` do not preserve energy in the system but my maths is very rusted.

At the time, I experimented with replacing Godot's Euler integration with Runge–Kutta fourth-order (which was, as per my research at the time, one of the most accurate integration method used in space sim).
That gets in fact rid of the wobbling, but no noticeable changes in precession so it didn't felt like a good improvement given the extra computational cost (again, when using sensible values for planets; I didn't really try blackhole-like object near the event horizon, although my planets where quite "dense").

Note, the tests were made for 2-body physics.

Overall, I would approve this PR, since it allows the user to get the expected result (stablish orbits) when using `gravity_distance_scale > 0`, even if it has limitations (in terms of level of simulation), it's better than the broken version we have now.

Faless reviewed (approved):
See my comment above.
But let's wait for another opinion from @madmiraal which is much more maths savvy than me.

ryanpeach commented:
Sorry about comment spamming, I tend to write in "train of thought"

madmiraal commented:
I think this PR needs a [proposal](https://github.com/godotengine/godot-proposals); so that it can receive a consensus from the community. The proposal should be about replacing the existing point gravity solution with a physically accurate solution for the reasons stated above.

This PR would then need to be updated in four ways:
1. The default `Gravity Distance Scale` should be 1 in 3D and 10 in 2D i.e. the distance from the `Gravity Point` at which gravity equals `Gravity`, and it needs to always be greater than 0 to avoid infinities. There shouldn't be a different solution at 0, which would be a cause of confusion.
2. Prevent division by 0, instead of checking whether or not a division by zero has taken place. The results of a division by zero are undefined, and are hardware and compiler dependent.
3. The `Math::Pow(x, 2)` should be replaced with `x * x`
4. The solution needs to be implemented for Bullet too; as previously requested.

ryanpeach commented:
I agree with @madmiraal and appreciate him taking the time to comment.

I'd like to ask some followup questions to his comment.

> The default Gravity Distance Scale should be 1 in 3D and 10 in 2D

Why 10? Just curious.

> it needs to always be greater than 0 to avoid infinities

I suggest using a log-scale slider then in the GUI

I don't currently like how godot's gui handles very small numbers.

> The Math::Pow(x, 2) should be replaced with x * x

I'm new to C++, what are the advantages?

> The solution needs to be implemented for Bullet too

It's in the physics3d server. Is bullet separate from that?

Further I'd like to make one contribution:

https://www.desmos.com/calculator/3bofinug4x

This is a desmos calculator comparing the different formula.

I show that gravity_distance_scale is actually an entirely unnecessary parameter, and since it's not used anywhere else in the code, personally I'd petition to simply remove it.

Using "gravity" alone, and putting it on a log slider, would be perfect IMO and avoid redundancy and 0's. Then make "point gravity" a checkbox. And we are done.

madmiraal commented:
> > The default Gravity Distance Scale should be 1 in 3D and 10 in 2D
>
> Why 10? Just curious.

The 2D constants are all 10x the 3D constants. For example the default gravity in 2D is 98 vs 3D 9.8. They just use two different arbitrary units.

> I suggest using a log-scale slider then in the GUI
>
> I don't currently like how godot's gui handles very small numbers.

AFAIK Godot doesn't have a log-scale slider (yet :wink:).

> > The Math::Pow(x, 2) should be replaced with x * x
>
> I'm new to C++, what are the advantages?

Speed.

> > The solution needs to be implemented for Bullet too
>
> It's in the physics3d server. Is bullet separate from that?

Yes, Godot has two 3D physics servers: Godot and Bullet.

> I show that gravity_distance_scale is actually an entirely unnecessary parameter once you remove the +1, and since it's not used anywhere else in the code, personally I'd petition to simply remove it.
>
> Using "gravity" alone, and putting it on a log slider, would be perfect IMO and avoid redundancy and 0's. Then make "point gravity" a checkbox. And we are done.

Good point. The `Gravity` of the `Gravity Point` is just G × M<sub>1</sub>, which we already have as 9.8 in 3D and 98 in 2D.

ryanpeach commented:
Please see the new PR. #42769

I'm providing this as an alternative.

Needs some more documentation changes and testing. But it builds on my machine.

ryanpeach commented:
I closed that PR. Gravity_distance_scale is useful for user convenience and computational reasons apparently.

Commit:
This else statement makes odd behavior in  practice for objects that are stationary on top of the gravity well (like attractors).

Faless commented:
When `gravity_distance_scale == 0` you must apply the fixed gravity `gravity += (p_area->get_transform().xform(p_area->get_gravity_vector()) - get_transform().get_origin()).normalized() * p_area->get_gravity();` (without taking distance into account).

Commit:
Fixed removal of graavity_distance_scale == 0 condition, while maintaining the previous commit issue

ryanpeach commented:
Sorry I didn't consider that part of the logic for a moment because it wasn't in my demo. Fixed.

Faless commented on servers/physics_2d/body_2d_sw.cpp:
```diff
@@ -384,11 +384,20 @@ void Body2DSW::set_space(Space2DSW *p_space) {
-			gravity += v.normalized() * (p_area->get_gravity() / Math::pow(v.length() * p_area->get_gravity_distance_scale() + 1, 2));
+		const real_t gravity_distance_scale = p_area->get_gravity_distance_scale();
+		Vector2 v = p_area->get_transform().xform(p_area->get_gravity_vector()) - get_transform().get_origin();
+		real_t v_length = v.length();
```
`v_length` only needs to be calculated if `gravity_distance_scale > 0`.

Commit:
v_length only needs to be calculated when gravity_distance_scale > 0

ryanpeach commented:
Good catch. Sorry for the messy mistakes.

pouleyKetchoupp commented on servers/physics_2d/body_2d_sw.cpp:
```diff
@@ -384,11 +384,20 @@ void Body2DSW::set_space(Space2DSW *p_space) {
+			real_t v_length = v.length();
+			if (v_length > 0) {
+				Vector2 new_gravity = gravity + (v.normalized() * (p_area->get_gravity() / Math::pow(v_length * gravity_distance_scale, 2)));
+				if (isinf(new_gravity.x) || isinf(new_gravity.y)) {
```
This check for infinity values is not needed anymore since you're already checking `v_length` and `gravity_distance_scale` are both > 0, so you can just remove the condition and replace with `gravity += ...`.

pouleyKetchoupp commented on servers/physics_2d/body_2d_sw.cpp:
```diff
@@ -384,11 +384,20 @@ void Body2DSW::set_space(Space2DSW *p_space) {
+		if (gravity_distance_scale > 0) {
+			real_t v_length = v.length();
+			if (v_length > 0) {
+				Vector2 new_gravity = gravity + (v.normalized() * (p_area->get_gravity() / Math::pow(v_length * gravity_distance_scale, 2)));
```
`Math::pow` could be replaced with a simple multiplication here for performance reasons. Most compilers would optimize it anyway, but I'm not sure it's the case for all supported compilers and platforms so better be safe and use the simpler code.

pouleyKetchoupp commented on servers/physics_3d/body_3d_sw.cpp:
```diff
@@ -398,11 +398,20 @@ void Body3DSW::set_space(Space3DSW *p_space) {
-		if (p_area->get_gravity_distance_scale() > 0) {
-			Vector3 v = p_area->get_transform().xform(p_area->get_gravity_vector()) - get_transform().get_origin();
-			gravity += v.normalized() * (p_area->get_gravity() / Math::pow(v.length() * p_area->get_gravity_distance_scale() + 1, 2));
+		const real_t gravity_distance_scale = p_area->get_gravity_distance_scale();
```
Same comments as in 2D.

pouleyKetchoupp reviewed (changes requested):
Sorry for the long delay! It looks great and we can definitely make this change for 4.0, I've just commented on some details.

Bullet is currently disabled on master, so there's no requirement to make the same changes in Bullet for now. I'll just make a tracker to keep a record of needed changes for later.

ryanpeach commented:
I'm no longer working on this. I hope my code is useful as is for someone to complete.

Mickeon commented:
Hasn't this already been salvaged by https://github.com/godotengine/godot/pull/52070?
